CMakeLists.txt

```diff
@@ -13,236 +13,280 @@
-	ELSE()
-		SET(HZ_BIT_FLAG " -m64 ")
-	ENDIF(${HZ_BIT} MATCHES "32")
+# find Threads
+find_package(Threads REQUIRED)
 
-	message(STATUS "${HZ_BIT} Bit")
+# find Boost
+find_package(Boost 1.72 REQUIRED COMPONENTS thread chrono)
 
-	set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -pipe -g -Wall -Werror ${HZ_BIT_FLAG} ${HZ_CODE_COVERAGE_COMPILE_FLAGS} ${HZ_VALGRIND_COMPILE_FLAGS}")
-	set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -pipe -O3 -Wall ${HZ_BIT_FLAG} ${HZ_CODE_COVERAGE_COMPILE_FLAGS} ${HZ_VALGRIND_COMPILE_FLAGS}")
+# find OpenSSL if building WITH_OPENSSL
+if (WITH_OPENSSL)
+    if (APPLE)
+        # This is a bug in CMake that causes it to prefer the system version over
+        # the one in the specified ROOT folder. see
+        # https://stackoverflow.com/questions/58446253/xcode-11-ld-error-your-binary-is-not-an-allowed-client-of-usr-lib-libcrypto-dy
+        set(OPENSSL_ROOT_DIR /usr/local/opt/openssl)
+        set(OPENSSL_INCLUDE_DIR ${OPENSSL_ROOT_DIR}/include CACHE FILEPATH "" FORCE)
+        set(OPENSSL_CRYPTO_LIBRARY ${OPENSSL_ROOT_DIR}/lib/libcrypto.dylib CACHE FILEPATH "" FORCE)
+        set(OPENSSL_SSL_LIBRARY ${OPENSSL_ROOT_DIR}/lib/libssl.dylib CACHE FILEPATH "" FORCE)
+    endif ()
 
-	add_library(${HZ_LIB_NAME} ${HZ_LIB_TYPE} ${HZ_SOURCES} ${INLINE_SOURCES} ${HZ_GENERATED_SOURCES} ${HZ_HEADERS} ${HZ_GENERATED_HEADERS})
+    find_package(OpenSSL REQUIRED)
+endif ()
 
-	link_libraries(${HZ_LIB_NAME} pthread rt)
-ENDIF(${CMAKE_SYSTEM_NAME} MATCHES "Linux")
 
-IF(${CMAKE_SYSTEM_NAME} MATCHES "Windows")
-	message(STATUS "WINDOWS ENVIRONMENT DETECTED ${CMAKE_GENERATOR} ${CMAKE_BUILD_TYPE} ")
-	add_definitions(-D__WIN${HZ_BIT}__ -DWIN${HZ_BIT} -D_WIN${HZ_BIT} -DNOMINMAX)
+function(add_hazelcast_library name type)
+    # add the library target
+    add_library(
+        ${name}
+        ${type}
+        ${SOURCE_FILES} ${HEADER_FILES}
+    )
+
+    # library requires c++11
+    target_compile_features(
+        ${name}
+        PUBLIC cxx_std_11
+    )
+
+    # MSVC-specific compiler flags
+    if (MSVC)
+        target_compile_options(${name} PRIVATE /bigobj)
+    endif ()
+
+    # links the library against the system's thread library
+    target_link_libraries(${name} PUBLIC Threads::Threads)
+
+    # add Boost::thread and Boost::chrono as dependencies
+    target_link_libraries(
+        ${name}
+        PUBLIC Boost::headers Boost::thread Boost::chrono
+    )
+    # set the Boost thread version
+    target_compile_definitions(
+        ${name}
+        PUBLIC BOOST_THREAD_VERSION=5
+    )
+
+    # If building WITH_OPENSSL, add OpenSSL::SSL and OpenSSL::Crypto as dependencies
+    # Both we and the user defines HZ_BUILD_WITH_SSL.
+    if (WITH_OPENSSL)
+        target_compile_definitions(
+            ${name}
+            PUBLIC HZ_BUILD_WITH_SSL
+        )
+        target_link_libraries(${name} PUBLIC OpenSSL::SSL OpenSSL::Crypto)
+    endif()
+
+    # windows-specific compile flags
+    if (WIN32)
```

Comment:
We used MSVC above, and the export flags may be only required when compiling with MSVC hence, this content may be added to the MSVC block above.

Reply:
I was also thinking about Windows compilers other than MSVC while writing this. (There are Intel and MinGW, I suppose).
For example, MinGW doesn't have the /bigobj option, but has the dllimport/export logic. 

Anyways, I am planning to use CMake's [GenerateExportHeader](https://cmake.org/cmake/help/latest/module/GenerateExportHeader.html), so that CMake can handle this import/export stuff for us. In that case, we won't need this block at all.